presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java

```diff
@@ -527,25 +530,26 @@ public boolean mkdirs(Path f, FsPermission permission)
         ListObjectsV2Request request = new ListObjectsV2Request()
                 .withBucketName(getBucketName(uri))
                 .withPrefix(key)
-                .withDelimiter(PATH_SEPARATOR)
                 .withRequesterPays(requesterPaysEnabled);
+        if (usePseudoDirectories) {
+            request.setDelimiter(PATH_SEPARATOR);
```

Comment:
i understand the conditional removal of `.setDelimiter(PATH_SEPARATOR)` as to list all sub directories in once shot, instead of hoping over dir levels. This is appropriate to directory listing but only for tables/partitions when `hive.recursive-directories` is on.

- shouldn’t this new option be enabled only if that option is enabled?
- shouldn’t we know that FS layer is called from the place where actually `hive.recursive-directories` matters — ie shouldn’t we modify the calling side to pass information that all subdirs are needed? Then we wouldn’t need the new toggle at all

cc @electrum

Reply:
To add to what @findepi said, I think with list caching done in `CachingDirectoryLister`, we can have an inconsistent (and sometimes incorrect) view of the list results in this approach. 
Further, doesn't lazy split generation done currently help in these scenarios?

Reply:
@findepi I reworked this to use hive.recursive-directories to control the behavior. Much cleaner this way.

@rohangarg - can you elaborate on when this could lead to an inconsistent list, and how thats different than walking the directories? I'm afraid I may be missing something there.

You're right that lazy split generation is good enough for most cases. The performance impact looks like it's significant when A) the number of subdirectories per partition is O(100), and B) the target query time is O(1 second). But in general, reducing the number of calls to S3 should be a good thing.

Reply:
The example that I was thinking where error can occur is : 
1. set recursive listing to true and query a path - with the feature, the full nested objects are cached in the list located status call (in CachingDirectoryLister's `cache` obj).
2. set recursive listing to false and query a path - now while listing the path, the nested objects should not occur in the result whereas the Lister's `cache` has complete nested listing stored.